hazelcast/src/main/java/com/hazelcast/map/impl/nearcache/NearCacheImpl.java

```diff
@@ -277,6 +207,79 @@ public NearCacheStatsImpl getNearCacheStats() {
+                            records.addAll(cache.values());
+                            int evictSize = (int) (cache.size() * EVICTION_FACTOR);
+                            int i = 0;
+                            for (NearCacheRecord record : records) {
+                                cache.remove(record.getKey());
+                                stats.incrementEvictions();
+                                updateSizeEstimator(-calculateCost(record));
+                                if (++i > evictSize) {
+                                    break;
+                                }
+                            }
+                        } finally {
+                            canEvict.set(true);
+                        }
+                        if (cache.size() >= maxSize && canEvict.compareAndSet(true, false)) {
+                            try {
+                                executionService.execute(NEAR_CACHE_EXECUTOR_NAME, this);
+                            } catch (RejectedExecutionException e) {
+                                canEvict.set(true);
+                            }
+                        }
+                    }
+                });
+            } catch (RejectedExecutionException e) {
+                canEvict.set(true);
+            } catch (Exception e) {
+                throw rethrow(e);
+            }
+        }
+    }
+
+    private void fireCacheExpiration() {
+        if (Clock.currentTimeMillis() < (lastExpiration + EXPIRATION_CLEANUP_INTERVAL_MILLIS)) {
+            return;
+        }
+
+        if (canExpire.compareAndSet(true, false)) {
+            try {
+                executionService.execute(NEAR_CACHE_EXECUTOR_NAME, new Runnable() {
+                    public void run() {
+                        try {
+                            lastExpiration = Clock.currentTimeMillis();
+                            for (Map.Entry<Data, NearCacheRecord> entry : cache.entrySet()) {
+                                if (entry.getValue().isExpired(maxIdleMillis, timeToLiveMillis)) {
+                                    Data key = entry.getKey();
+                                    NearCacheRecord record = cache.remove(key);
+                                    // if a mapping exists
+                                    if (record != null) {
+                                        updateSizeEstimator(-calculateCost(record));
+                                        stats.incrementExpirations();
+                                    }
+                                }
+                            }
+                        } finally {
+                            canExpire.set(true);
+                        }
+                    }
+                });
+            } catch (RejectedExecutionException e) {
+                canExpire.set(true);
+            } catch (Exception e) {
+                throw rethrow(e);
```

Comment:
You are potentially left with a locked canExpire if an exception is thrown. 

I normally do something like this:
```
boolean unlock = true;
try{
     someaction
    unlock = false
}finally{
   if(unlock) doUnlock()
}
```

It matches the locking idiom so you always know that something gets done.